### PR TITLE
Add mkwikitable

### PR DESCRIPTION
**Pull request by kerel-fs**
This script generates a mediawiki table about all ipv4/ipv6-networks.
See https://wiki.freifunk.net/IP-Netze/IC-VPN

Comments appreciated.

I am using the following to publish the generated data:
```bash
./mkwikitable | xclip -selection c
# Now C&P into the Freifunk-Wiki.
```

### CODE REVIEW

**Comment by mweinelt** [comment]
How is the "WIP" going?

File: apireader.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import json
+from argparse import ArgumentParser
+
+DUMP_NAME = "api_data.dump"
+
+
+def get_api_dict(src_dir):
+    with open(src_dir + "directory.json", 'r') as directory_file:
+        api_dict = json.load(directory_file)
+    return api_dict
```
Comment (mweinelt) [inline apireader.py]: How about loading this directly from github and not depending on having that repository around? I mean it's just one file.

```
def get_api_dict(src_dir):
    uri = "https://raw.githubusercontent.com/freifunk/directory.api.freifunk.net/master/directory.json"
    response = requests.get(uri)
    api_dict = json.loads(response.text)
    return api_dict
```

**Comment by mweinelt** [comment]
Please check if icvpn-meta exists _before_ download all the api files.

File: apireader.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def get_api_data_from_git(api_dict):
+    data = dict()
+    for community in api_dict:
+        api_url = api_dict[community]
+        print("Loading %s..." % api_url, file=sys.stderr)
```
Comment (mweinelt) [inline apireader.py]: Don't use string interpolation, instead use string formatting. Printing the community would also be useful...

```
        print("Loading {:<16} {}...".format(community, api_url), file=sys.stderr)
```

File: apireader.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        try:
+            response = requests.get(api_url, timeout=2)
+            community_api = response.json()
+            data[community] = community_api
+        except requests.exceptions.RequestException:
+            print("Error: Couldn't load %s." % api_url, file=sys.stderr)
```
Comment (mweinelt) [inline apireader.py]: ...especially on errors, which we can report back to the communities.

File: apireader.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            continue
+        except ValueError:
+            print("Error: Couldn't parse %s." % api_url, file=sys.stderr)
```
Comment (mweinelt) [inline apireader.py]: Again, same as above.

File: mkwikitable [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+def get_networklist(meta_src, api_src, family):
+    """
+    Returns a sorted list of all freifunk networks (family:ipv4/ipv6).
+    Each entry is a dict containing the network and
+    additional information about the network (asn, tech-c, community_name and community_url).
+    """
+    all_networks = []
+
+    api_dict = get_api_dict(api_src)
+    api_data = get_api_data(api_dict)
```
Comment (mweinelt) [inline mkwikitable]: You are currently fetching the metadata twice, for ipv4 and ipv6. This should be reduced to just once.

**Comment by kerel-fs** [comment]
Danke @mweinelt für die Kommentare! Ich selbst werde in den nächsten Tagen leider nicht dazu kommen das fertigzustellen.

File: README.md [hunk 6]
```diff
@@ -206,6 +206,15 @@ Options:
 [Smokeping]: http://oss.oetiker.ch/smokeping/
 
 
+Wiki ('mkwikitable')
+--------------------
+`mkwikitable` generates ...
+
+The output of `mkwikitable --help`:
+```
```
Comment (mweinelt) [inline README.md]: missing help text

File: apireader.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        try:
+            response = requests.get(api_url, timeout=2)
+            community_api = response.json()
+            data[community] = community_api
+        except requests.exceptions.RequestException:
+            print("Error: Couldn't load api file for {} from {}.".format(community, api_url), file=sys.stderr)
+            continue
+        except ValueError:
+            print("Error: Couldn't parse api file for {} from {}.".format(community, api_url), file=sys.stderr)
+            continue
```
Comment (mweinelt) [inline apireader.py]: when handling both exceptions the same way, group them into a single except-clause.

```
except (requests.exceptions.RequestException, ValueError):
```

File: apireader.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def get_api_data_from_file(filename):
+    with open(filename, 'r') as dump_file:
+        api_data = json.load(dump_file)
+    return api_data
+
+
+def get_api_data(api_dict):
+    # NOTE:Unexpected behavior!!
```
Comment (mweinelt) [inline apireader.py]: What unexpected behaviour?

File: mkwikitable [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import sys
+import os.path
+import jinja2
+import json
+from argparse import ArgumentParser
+from socket import AF_INET, AF_INET6, inet_pton
+from filereader import get_communities_data
+from apireader import get_api_dict, get_api_data
+
+templates = {"ip-netze.j2": """====IPv6====
```
Comment (mweinelt) [inline mkwikitable]: `template` actually.

File: mkwikitable [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+{%- endfor -%}
+{%- endfor %}
+|- bgcolor="#efefef"
+! Netzwerk
+! Community
+! AS
+! Kontakt (tech-c)
+|}"""}
+
+# templates = {"ip-netze.j2":"""
```
Comment (mweinelt) [inline mkwikitable]: is this still necessary/useful?

File: mkwikitable [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    """
+    all_networks = []
+
+    for community, data in get_communities_data(srcdir=meta_src, exclude=""):
+        try:
+            networks = data["networks"][family]
+        except (TypeError, KeyError):  # NOTE: no networks configured
+            continue
+
+        try:
```
Comment (mweinelt) [inline mkwikitable]: asn = data.get("asn", "")

File: mkwikitable [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            asn = data["asn"]
+        except KeyError:
+            asn = ""
+
+        try:
```
Comment (mweinelt) [inline mkwikitable]: same as above

File: mkwikitable [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            techc = list(data["tech-c"])
+        except (TypeError, KeyError):
+            techc = list()
+
+        try:
```
Comment (mweinelt) [inline mkwikitable]: same as above

File: mkwikitable [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+            community_url = api_data[community]["url"]
+        except KeyError:
+            community_url = ""
+
+        try:
```
Comment (mweinelt) [inline mkwikitable]: same as above

File: apireader.py (hunk 8 above)
Comment (kerel-fs) [inline apireader.py]: Reading from "api_data.dump" instead of loading the actual data is not documented.

File: mkwikitable [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+    adr_family = AF_INET6 if family == "ipv6" else AF_INET
+    all_networks = sorted(all_networks, key=lambda x:
+                          inet_pton(adr_family, x["network"].split("/")[0]))
+    return all_networks
+
+
+def mkwikitable(meta_src):
+    if not os.path.isdir(meta_src):
+        print("Error: Couldn't find icvpn-meta directory {}.".format(meta_src), file=sys.stderr)
```
Comment (mweinelt) [inline mkwikitable]: `directory` => `repository`

File: apireader.py (hunk 8 above)
Comment (mweinelt) [inline apireader.py]: So this should be marked as private instead, I guess? Usually you would prefix underbars.

**Pull request merged**